Applies variable size transfer to 'ParticleHandler'.

marcfehling opened the pull request:
This pull requests implements the new variable size transfer functionality of p4est in the `ParticleHandler` class.

The `store_particles` function takes the `std::multimap` apart and extracts all particles belonging to a cell, whereas the `load_particles` function adds these particles with the correct level back in the map.

As mentioned in #6790, there are other functions dedicated to sending data in the `ParticleHandler` class, but none of them use the interface to `p4est`. Thus, my initial concern of having to change them too turned out to be wrong. I felt free to close #6790 as soon as this pull request is merged.

@gassmoeller -- I do not have a suitable example program in parallel at hand that uses excessive amounts of particles. Could you try this branch out to run a small performance test, to check it these changes have any impact on it?

@bangerth  -- FYI

gassmoeller commented:
Unfortunately I have no big test case either at the moment. I never finished geodynamics/aspect#1957, which would finally use the deal.II functionality instead of the duplicated code in aspect. I guess we can only rely on the deal.II tests to test for correctness and need to test the speed later.

bangerth commented on source/particles/particle_handler.cc:
```diff
@@ -1118,101 +1118,84 @@ namespace Particles
+                          particles_in_cell.second,
+                          [&stored_particles_on_cell](
+                            std::pair<internal::LevelInd,
+                                      Particle<dim, spacedim>> particle) {
```
make this argument a const reference

bangerth commented on source/particles/particle_handler.cc:
```diff
@@ -1118,101 +1118,84 @@ namespace Particles
+                              particles_in_cell.second,
+                              [&stored_particles_on_cell](
+                                std::pair<internal::LevelInd,
+                                          Particle<dim, spacedim>> particle) {
```
same here

bangerth commented on source/particles/particle_handler.cc:
```diff
@@ -1118,101 +1118,84 @@ namespace Particles
-      Assert(false, ExcInternalError());
 
-    return data_vector;
+    Assert(n_particles == stored_particles_on_cell.size(), ExcInternalError());
```
move this assertion right after the two places where you call `std::foreach` to copy the particles. If you do this, then you can also make the `n_particles` variable a local variable in the two code blocks where you use it, rather than making it a function-level variable.

bangerth commented on source/particles/particle_handler.cc:
```diff
@@ -1222,140 +1205,141 @@ namespace Particles
-
-    if (*n_particles_in_cell_ptr == 0)
-      return;
+    std::vector<Particle<dim, spacedim>> loaded_particles_on_cell =
```
I had been wondering why you didn't make this variable `const` until I realized that you (cleverly!) use `std::move` when building the multimap again. Maybe add a comment here to this effect.

bangerth commented on source/particles/particle_handler.cc:
```diff
@@ -1222,140 +1205,141 @@ namespace Particles
+                                                  std::move(particle)));
 #  endif
-            ++position_hint;
+                ++position_hint;
```
add a comment here like "move the hint position forward by one, i.e., for the next particle, the 'hint' position will be right after the one just inserted" or similar

marcfehling commented:
Ready for tests and merge once done.

masterleinad commented:
`Applies bangerth's suggestions. (1)` is probably not a good commit message. :wink: Before merging you should merge this commit into the other ones.

masterleinad commented:
/run-tests

marcfehling commented:
I usually squash all commits when we reach the testing phase of a pull request, i.e. when someone triggers it. Will do know ;-)

jppelteret commented:
@bangerth You need to mark this as approved before it could be merged.

masterleinad commented:
Locally, there is one test failing:
```
particles/particle_handler_05.debug: RUN failed. ------ Partial output:
4309: JobId simserv02 Fri Aug  3 19:39:51 2018
4309: DEAL:0:2d/2d::Before refinement particle id 0 is in cell 0.0
4309: DEAL:0:2d/2d::Before refinement particle id 1 is in cell 0.0
4309: DEAL:0:2d/2d::Before refinement particle id 2 is in cell 0.0
4309: DEAL:0:2d/2d::Before refinement particle id 3 is in cell 0.0
4309: DEAL:0:2d/2d::Before refinement particle id 4 is in cell 0.0
4309: DEAL:0:2d/2d::Before refinement particle id 5 is in cell 0.0
4309: DEAL:0:2d/2d::Before refinement particle id 6 is in cell 0.0
4309: DEAL:0:2d/2d::Before refinement particle id 7 is in cell 0.0
4309: DEAL:0:2d/2d::Before refinement particle id 8 is in cell 0.0
4309: DEAL:0:2d/2d::After refinement particle id 0 is in cell 1.0
4309: DEAL:0:2d/2d::After refinement particle id 1 is in cell 1.0
4309: DEAL:0:2d/2d::After refinement particle id 3 is in cell 1.0
4309: DEAL:0:2d/2d::After refinement particle id 4 is in cell 1.0
4309: DEAL:0:2d/2d::After refinement particle id 6 is in cell 1.1
4309: DEAL:0:2d/2d::After refinement particle id 7 is in cell 1.1
4309: DEAL:0:2d/2d::After refinement particle id 2 is in cell 1.2
4309: DEAL:0:2d/2d::After refinement particle id 5 is in cell 1.2
4309: DEAL:0:2d/2d::After refinement particle id 8 is in cell 1.3
4309: 
4309: particles/particle_handler_05.debug: RUN failed. ------ Additional output on stdout/stderr:
4309: 
4309: 
4309: --------------------------------------------------------
4309: An error occurred in line <1194> of file <../source/particles/particle_handler.cc> in function
4309:     std::vector<char, std::allocator<char> > dealii::Particles::ParticleHandler<<anonymous>, <anonymous> >::store_particles(const typename dealii::Triangulation<dim, spacedim>::cell_iterator&, typename dealii::Triangulation<dim, spacedim>::CellStatus) const [with int dim = 2; int spacedim = 2; typename dealii::Triangulation<dim, spacedim>::cell_iterator = dealii::TriaIterator<dealii::CellAccessor<2, 2> >; typename dealii::Triangulation<dim, spacedim>::CellStatus = dealii::Triangulation<2, 2>::CellStatus]
4309: The violated condition was: 
4309:     (n_particles) == (stored_particles_on_cell.size())
4309: Additional information: 
4309:     Dimension 9 not equal to 4.
4309: --------------------------------------------------------
4309: 
4309: [simserv02:103049] *** Process received signal ***
4309: [simserv02:103049] Signal: Aborted (6)
4309: [simserv02:103049] Signal code:  (-6)
4309: [simserv02:103049] [ 0] /lib/x86_64-linux-gnu/libpthread.so.0(+0x11390)[0x7f71c526b390]
4309: [simserv02:103049] [ 1] /lib/x86_64-linux-gnu/libc.so.6(gsignal+0x38)[0x7f71c3b57428]
4309: [simserv02:103049] [ 2] /lib/x86_64-linux-gnu/libc.so.6(abort+0x16a)[0x7f71c3b5902a]
4309: [simserv02:103049] [ 3] /mnt/data/darndt/Sources/dealii-gcc-8/build/lib/libdeal_II.g.so.9.1.0-pre(_ZN6dealii18deal_II_exceptions9internals22do_issue_error_nothrowERKNS_13ExceptionBaseE+0x0)[0x7f71f1966011]
4309: [simserv02:103049] [ 4] /mnt/data/darndt/Sources/dealii-gcc-8/build/lib/libdeal_II.g.so.9.1.0-pre(_ZN6dealii18deal_II_exceptions9internals5abortERKNS_13ExceptionBaseE+0x29)[0x7f71f19660fc]
4309: [simserv02:103049] [ 5] /mnt/data/darndt/Sources/dealii-gcc-8/build/lib/libdeal_II.g.so.9.1.0-pre(_ZN6dealii18deal_II_exceptions9internals20issue_error_noreturnINS_18StandardExceptions20ExcDimensionMismatchEEEvNS1_17ExceptionHandlingEPKciS7_S7_S7_T_+0x60)[0x7f71eab823da]
4309: [simserv02:103049] [ 6] /mnt/data/darndt/Sources/dealii-gcc-8/build/lib/libdeal_II.g.so.9.1.0-pre(_ZNK6dealii9Particles15ParticleHandlerILi2ELi2EE15store_particlesERKNS_12TriaIteratorINS_12CellAccessorILi2ELi2EEEEENS_13TriangulationILi2ELi2EE10CellStatusE+0x4b1)[0x7f71f33d8edb]
4309: [simserv02:103049] [ 7] /mnt/data/darndt/Sources/dealii-gcc-8/build/lib/libdeal_II.g.so.9.1.0-pre(_ZSt13__invoke_implISt6vectorIcSaIcEERMN6dealii9Particles15ParticleHandlerILi2ELi2EEEKFS2_RKNS3_12TriaIteratorINS3_12CellAccessorILi2ELi2EEEEENS3_13TriangulationILi2ELi2EE10CellStatusEERKS6_JSC_SF_EET_St19__invoke_memfun_refOT0_OT1_DpOT2_+0x9d)[0x7f71f343080c]
4309: [simserv02:103049] [ 8] /mnt/data/darndt/Sources/dealii-gcc-8/build/lib/libdeal_II.g.so.9.1.0-pre(_ZSt8__invokeIRMN6dealii9Particles15ParticleHandlerILi2ELi2EEEKFSt6vectorIcSaIcEERKNS0_12TriaIteratorINS0_12CellAccessorILi2ELi2EEEEENS0_13TriangulationILi2ELi2EE10CellStatusEEJRKS3_SC_SF_EENSt15__invoke_resultIT_JDpT0_EE4typeEOSM_DpOSN_+0x72)[0x7f71f342af8c]
4309: [simserv02:103049] [ 9] /mnt/data/darndt/Sources/dealii-gcc-8/build/lib/libdeal_II.g.so.9.1.0-pre(_ZNSt5_BindIFMN6dealii9Particles15ParticleHandlerILi2ELi2EEEKFSt6vectorIcSaIcEERKNS0_12TriaIteratorINS0_12CellAccessorILi2ELi2EEEEENS0_13TriangulationILi2ELi2EE10CellStatusEESt17reference_wrapperIKS3_ESt12_PlaceholderILi1EESL_ILi2EEEE6__callIS6_JSC_OSF_EJLm0ELm1ELm2EEEET_OSt5tupleIJDpT0_EESt12_Index_tupleIJXspT1_EEE+0xa8)[0x7f71f3421aac]
4309: [simserv02:103049] [10] /mnt/data/darndt/Sources/dealii-gcc-8/build/lib/libdeal_II.g.so.9.1.0-pre(_ZNSt5_BindIFMN6dealii9Particles15ParticleHandlerILi2ELi2EEEKFSt6vectorIcSaIcEERKNS0_12TriaIteratorINS0_12CellAccessorILi2ELi2EEEEENS0_13TriangulationILi2ELi2EE10CellStatusEESt17reference_wrapperIKS3_ESt12_PlaceholderILi1EESL_ILi2EEEEclIJSC_SF_ES6_EET0_DpOT_+0x60)[0x7f71f34120c8]
4309: [simserv02:103049] [11] /mnt/data/darndt/Sources/dealii-gcc-8/build/lib/libdeal_II.g.so.9.1.0-pre(_ZNSt17_Function_handlerIFSt6vectorIcSaIcEERKN6dealii12TriaIteratorINS3_12CellAccessorILi2ELi2EEEEENS3_13TriangulationILi2ELi2EE10CellStatusEESt5_BindIFMNS3_9Particles15ParticleHandlerILi2ELi2EEEKFS2_S9_SC_ESt17reference_wrapperIKSH_ESt12_PlaceholderILi1EESN_ILi2EEEEE9_M_invokeERKSt9_Any_dataS9_OSC_+0x5a)[0x7f71f33feb6e]
4309: [simserv02:103049] [12] /mnt/data/darndt/Sources/dealii-gcc-8/build/lib/libdeal_II.g.so.9.1.0-pre(_ZNKSt8functionIFSt6vectorIcSaIcEERKN6dealii12TriaIteratorINS3_12CellAccessorILi2ELi2EEEEENS3_13TriangulationILi2ELi2EE10CellStatusEEEclES9_SC_+0x65)[0x7f71f2e424af]
4309: [simserv02:103049] [13] /mnt/data/darndt/Sources/dealii-gcc-8/build/lib/libdeal_II.g.so.9.1.0-pre(_ZNSt17_Function_handlerIFSt6vectorIcSaIcEEN6dealii12TriaIteratorINS3_12CellAccessorILi2ELi2EEEEENS3_13TriangulationILi2ELi2EE10CellStatusEESt8functionIFS2_RKS7_SA_EEE9_M_invokeERKSt9_Any_dataOS7_OSA_+0x58)[0x7f71f2e3d926]
4309: [simserv02:103049] [14] /mnt/data/darndt/Sources/dealii-gcc-8/build/lib/libdeal_II.g.so.9.1.0-pre(_ZNKSt8functionIFSt6vectorIcSaIcEEN6dealii12TriaIteratorINS3_12CellAccessorILi2ELi2EEEEENS3_13TriangulationILi2ELi2EE10CellStatusEEEclES7_SA_+0x65)[0x7f71f2e37d4f]
4309: [simserv02:103049] [15] /mnt/data/darndt/Sources/dealii-gcc-8/build/lib/libdeal_II.g.so.9.1.0-pre(_ZN6dealii8parallel11distributed13TriangulationILi2ELi2EE12DataTransfer9pack_dataERKSt6vectorISt5tupleIJP14p4est_quadrantNS_13TriangulationILi2ELi2EE10CellStatusENS_12TriaIteratorINS_12CellAccessorILi2ELi2EEEEEEESaISG_EERKS5_ISt8functionIFS5_IcSaIcEESF_SB_EESaISP_EEST_+0x8eb)[0x7f71f2e178f9]
4309: [simserv02:103049] [16] /mnt/data/darndt/Sources/dealii-gcc-8/build/lib/libdeal_II.g.so.9.1.0-pre(_ZN6dealii8parallel11distributed13TriangulationILi2ELi2EE33execute_coarsening_and_refinementEv+0x6c8)[0x7f71f2e10606]
4309: [simserv02:103049] [17] /mnt/data/darndt/Sources/dealii-gcc-8/build/tests/particles/particle_handler_05.debug/particle_handler_05.debug(_Z4testILi2ELi2EEvv+0x4be)[0x45d51d]
4309: [simserv02:103049] [18] /mnt/data/darndt/Sources/dealii-gcc-8/build/tests/particles/particle_handler_05.debug/particle_handler_05.debug(main+0xae)[0x455da1]
4309: [simserv02:103049] [19] /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf0)[0x7f71c3b42830]
4309: [simserv02:103049] [20] /mnt/data/darndt/Sources/dealii-gcc-8/build/tests/particles/particle_handler_05.debug/particle_handler_05.debug(_start+0x29)[0x454389]
4309: [simserv02:103049] *** End of error message ***
4309: Aborted
4309: CMakeFiles/particle_handler_05.debug.diff.dir/build.make:69: recipe for target 'particle_handler_05.debug/output' failed
4309: CMakeFiles/Makefile2:291: recipe for target 'CMakeFiles/particle_handler_05.debug.diff.dir/all' failed
4309: CMakeFiles/Makefile2:298: recipe for target 'CMakeFiles/particle_handler_05.debug.diff.dir/rule' failed
4309: Makefile:220: recipe for target 'particle_handler_05.debug.diff' failed
4309: 
4309: 
4309: particles/particle_handler_05.debug: ******    RUN failed    *******
4309: 
4309: ===============================    OUTPUT END   ===============================
4309: Expected stage PASSED - aborting
4309: CMake Error at /mnt/data/darndt/Sources/dealii-gcc-8/build/share/deal.II/scripts/run_test.cmake:140 (MESSAGE):
4309:   *** abort
```
@marcfehling Can you confirm that problem?

masterleinad commented:
I guess this should simply be
```
diff --git a/source/particles/particle_handler.cc b/source/particles/particle_handler.cc
index 43f2151..0da395b 100644
--- a/source/particles/particle_handler.cc
+++ b/source/particles/particle_handler.cc
@@ -1190,10 +1190,9 @@ namespace Particles
                       &particle) {
                     stored_particles_on_cell.push_back(particle.second);
                   });
-
-                Assert(n_particles == stored_particles_on_cell.size(),
-                       ExcInternalError());
               }
+
+            AssertDimension(n_particles, stored_particles_on_cell.size());
           }
           break;
```

masterleinad commented on source/particles/particle_handler.cc:
```diff
@@ -1118,101 +1118,92 @@ namespace Particles
+                  });
+
+                Assert(n_particles == stored_particles_on_cell.size(),
+                       ExcInternalError());
```
This has to be outside the loop.

marcfehling replied:
Woops. My bad :(

masterleinad commented:
You just changed that one line, didn't you?

marcfehling commented:
There are two assertions in `store_particles`. I changed both to `AssertDimension`, and moved the first one behind the loop.

I was too eager squashing the commits into one. Sorry for that.
